### ISSUE TITLE
[STKAudioPlayer dealloc] doesn’t work

### ISSUE THREAD

**Issue by QueenaHuang** [tumtumtum/StreamingKit#323]
#144 

I recreate STKAudioPlauyer when player has unexpectedError. 
But [STKAudioPlayer dealloc] doesn’t work and it has memory leaks.

`/// Raised when an unexpected and possibly unrecoverable error has occured (usually best to recreate the STKAudioPlauyer)
-(void) audioPlayer:(STKAudioPlayer*)audioPlayer unexpectedError:(STKAudioPlayerErrorCode)errorCode;`

It's because that STKAudioPlauyer has singleton instance? 
So STKAudioPlauyer never call dealloc method?

It’s my code:

`@property (nonatomic, strong) STKAudioPlayer *player;
-(void) audioPlayer:(STKAudioPlayer *)audioPlayer unexpectedError:(STKAudioPlayerErrorCode)errorCode
{
    [self.player dispose];

    self.player = [[STKAudioPlayer alloc] initWithOptions:(STKAudioPlayerOptions){
        .bufferSizeInSeconds = 300,
        .secondsRequiredToStartPlaying = 300,
        .secondsRequiredToStartPlayingAfterBufferUnderun = 300}];
    self.player.meteringEnabled = YES;
    self.player.delegate = self;
}`

**Comment by wendongsheng** [comment]
your problem is solved？I have a same problem

**Comment by doriansgithub** [comment]
We have that blank and have no issues.

**Comment by iDevelopper** [comment]
If you want dealloc method to be called, remove the frame filters entry:

```
@property (strong, nonatomic) STKAudioPlayer *audioPlayer;

...

    self.audioPlayer = [[STKAudioPlayer alloc] initWithOptions:(STKAudioPlayerOptions){ .flushQueueOnSeek = YES, .enableVolumeMixer = YES, .secondsRequiredToStartPlaying = 0.001, .equalizerBandFrequencies = {32, 42, 52, 64, 74, 14000, 16000, 0} }];
    
    _audioPlayer.delegate = self;
    _audioPlayer.meteringEnabled = YES;
    _audioPlayer.equalizerEnabled = YES;
    _audioPlayer.volume = 1;

    [_audioPlayer appendFrameFilterWithName:@"CustomEQFilter" block:^(UInt32 channelsPerFrame, UInt32 bytesPerFrame, UInt32 frameCount, void* frames)
     {
         // ....
     }];

...
...

    // Dispose
    NSArray *frameFilters = [_audioPlayer frameFilters];
    
    for (STKFrameFilterEntry *entry in frameFilters) {
        [_audioPlayer removeFrameFilterWithName:entry.name];
    }
    
    [_audioPlayer dispose];
```

**Comment by wendongsheng** [comment]
Thank you very much


发自网易邮箱 iPhone 版
On 06/18/2017 17:02, Patrick wrote:

If you want dealloc method to be called, remove the frame filters entry:

@property (strong, nonatomic) STKAudioPlayer *audioPlayer;

...

    self.audioPlayer = [[STKAudioPlayer alloc] initWithOptions:(STKAudioPlayerOptions){ .flushQueueOnSeek = YES, .enableVolumeMixer = YES, .secondsRequiredToStartPlaying = 0.001, .equalizerBandFrequencies = {32, 42, 52, 64, 74, 14000, 16000, 0} }];
    
    _audioPlayer.delegate = self;
    _audioPlayer.meteringEnabled = YES;
    _audioPlayer.equalizerEnabled = YES;
    _audioPlayer.volume = 1;

    [_audioPlayer appendFrameFilterWithName:@"CustomEQFilter" block:^(UInt32 channelsPerFrame, UInt32 bytesPerFrame, UInt32 frameCount, void* frames)
     {
         // ....
     }];

...
...

    // Dispose
    NSArray *frameFilters = [_audioPlayer frameFilters];
    
    for (STKFrameFilterEntry *entry in frameFilters) {
        [_audioPlayer removeFrameFilterWithName:entry.name];
    }
    
    [_audioPlayer dispose];

**Comment by haritowa** [comment]
I believe I have solved your issue: https://github.com/tumtumtum/StreamingKit/pull/391

**Comment by diegostamigni** [comment]
Closed as mentioned #391 has been merged in.
